doc/dev/sharp_bits.rst

```diff
@@ -1049,12 +1049,41 @@ conditional statements, are also supported:
 >>> f(5)
 array([21., 21., 21., 21., 21.])
 
-However, capturing dynamic-shaped arrays within control-flow from outer scopes is currently not
-supported:
+By default, Catalyst for loops and while loops will automatically capture
+dynamically-shaped arrays from outside their scope for use within the loop,
+however loops cannot modify (i.e., return) array shapes and sizes:
+
+>>> @qjit()
+... def f(N):
+...     a = jnp.ones([N], dtype=float)
+...     @for_loop(0, 10, 1)
+...     def loop(i, _):
+...         return jnp.ones([i], dtype=float) # return array of new dimensions
+...     return loop(a)
+>>> f(5)
+AssertionError:
+result_types=[RankedTensorType(tensor<?xf64>)] doesn't match
+jax_ctx.avals_out=[ShapedArray(int64[], weak_type=True), f64[c]]
+
+In order to support modifying of array shape, size or dimension *within* a loop,
+the ``allow_array_resizing`` argument can be used:
```

Comment:
changing shape == adding new axes? Unless I am mistaken, isn't this always supported via things like `jnp.stack`?

Reply:
But it might not be supported across loop iterations. The type still has to be identical in the input and output of a loop-carried value, the only reason we can get away with changing the "size" is because if a dimension is unknown it could be 2 in the input and 5 in the output, but the type is the same with the unknown dimension.

Reply:
Had a discussion in real life, it turns out that there are several bugs involved in attempting to reshape dynamic shaped arrays. In particular:

- `jnp.reshape` of a dynamically-shaped array with new compile-time shape results in the array being returned without any change in shape:

  ```python
  @qjit(abstracted_axes={0: 'm', 1: 'n'}, keep_intermediate=True)
  def g(a):
       return jnp.reshape(a, (3, 5))
  
  a = jnp.ones([1, 3], dtype=float)
  g(a)
  ```

  I would expect the output to be of shape `(3, 5)` but with junk values for the out of bound elements.

- `jnp.reshape` of a dynamically-shaped array with new dynamic shape results in a segfault during compilation. Lowering to MLIR seems to happen correctly, but we suspect that `mhlo.dynamic_reshape` doesn't have a lowering rule.

  ```python
  @qjit(abstracted_axes={0: 'm', 1: 'n'}, keep_intermediate=True)
  def g(a):
       return jnp.reshape(a, (a.shape[1], a.shape[0]))
  
  a = jnp.ones([1, 3], dtype=float)
  g(a)
  ```

  - Separately, we should consider adding a proper exception to catch this, rather than the segfault killing the kernel.

- `jnp.reshape` of a dynamically-shaped array within a loop iteration leads to a cryptic error even before compilation:

  ```python
  >>> @qjit(abstracted_axes={0: 'm', 1: 'n'})
  ... def g(x):
  ...     @catalyst.for_loop(0, 10, 1, experimental_preserve_dimensions=False)
  ...     def loop(_, a):
  ...         return jnp.reshape(a, (3, 1))
  ...     return loop(x)
  >>> a = jnp.ones([1, 3], dtype=float)
  >>> g(a)
  ValueError: Too few leaves for PyTreeDef; expected 1, got 0
  ```